pkg/bindinfo/binding_auto.go

```diff
@@ -0,0 +1,150 @@
+	"github.com/pingcap/tidb/pkg/util/logutil"
+	utilparser "github.com/pingcap/tidb/pkg/util/parser"
+	"go.uber.org/zap"
+)
+
+// BindingPlanInfo represents the binding plan info.
+type BindingPlanInfo struct {
+	*Binding
+
+	// Info from StmtStats
+	Plan                 string
+	AvgLatency           float64
+	ExecTimes            int64
+	AvgScanRows          float64
+	AvgReturnedRows      float64
+	LatencyPerReturnRow  float64
+	ScanRowsPerReturnRow float64
+
+	// Recommendation
+	Recommend string
+	Reason    string
+}
+
+// BindingAuto represents a series of APIs that help manage bindings automatically.
+type BindingAuto interface {
+	// TODO: RecordHistPlansAsBindings records the history plans as bindings for qualified queries.
+
+	// ShowPlansForSQL shows historical plans for a specific SQL.
+	ShowPlansForSQL(currentDB, sqlOrDigest, charset, collation string) ([]*BindingPlanInfo, error)
+}
+
+type bindingAuto struct {
+	sPool util.DestroyableSessionPool
+}
+
+func newBindingAuto(sPool util.DestroyableSessionPool) BindingAuto {
+	return &bindingAuto{
+		sPool: sPool,
+	}
+}
+
+// ShowPlansForSQL shows historical plans for a specific SQL.
+func (ba *bindingAuto) ShowPlansForSQL(currentDB, sqlOrDigest, charset, collation string) ([]*BindingPlanInfo, error) {
+	// parse and normalize sqlOrDigest
+	p := parser.New()
+	var normalizedSQL, whereCond string
+	stmtNode, err := p.ParseOneStmt(sqlOrDigest, charset, collation)
+	if err == nil {
+		db := utilparser.GetDefaultDB(stmtNode, currentDB)
+		normalizedSQL, _ = NormalizeStmtForBinding(stmtNode, db, false)
+	}
+
+	// read binding data from mysql.bind_info
+	if normalizedSQL != "" {
+		whereCond = fmt.Sprintf("where original_sql='%s'", normalizedSQL)
+	} else { // treat sqlOrDigest as a digest
+		whereCond = fmt.Sprintf("where sql_digest='%s'", sqlOrDigest)
+	}
+	bindings, err := readBindingsFromStorage(ba.sPool, whereCond)
+	if err != nil {
+		return nil, err
+	}
+
+	// read plan info from information_schema.tidb_statements_stats
+	bindingPlans := make([]*BindingPlanInfo, 0, len(bindings))
+	for _, binding := range bindings {
+		pInfo, err := ba.getPlanExecInfo(binding.PlanDigest)
+		if err != nil {
+			logutil.BgLogger().Error("getStmtStatsByDigestInCluster", zap.String("plan_digest", binding.PlanDigest), zap.Error(err))
+			continue
+		}
+		autoBinding := &BindingPlanInfo{Binding: binding}
+		if pInfo != nil && pInfo.ExecCount > 0 {
+			autoBinding.Plan = pInfo.Plan
+			autoBinding.ExecTimes = pInfo.ExecCount
+			autoBinding.AvgLatency = float64(pInfo.TotalTime) / float64(pInfo.ExecCount)
+			autoBinding.AvgScanRows = float64(pInfo.ProcessedKeys) / float64(pInfo.ExecCount)
+			autoBinding.AvgReturnedRows = float64(pInfo.ResultRows) / float64(pInfo.ExecCount)
+			autoBinding.LatencyPerReturnRow = autoBinding.AvgLatency / autoBinding.AvgReturnedRows
+			autoBinding.ScanRowsPerReturnRow = autoBinding.AvgScanRows / autoBinding.AvgReturnedRows
+		}
+		bindingPlans = append(bindingPlans, autoBinding)
```

Comment:
If `pInfo` is `nil`, do we still need to do this?

Reply:
Yes, then in this case, we can still see the binding, but all plan execution info is empty. I added some comments here.